src/abstractdataframe/unique.jl

```diff
@@ -0,0 +1,392 @@
+- `cols` : a selector specifying the column(s) or their transformations to
+  compare. Can be any column selector or transformation accepted by
+  [`select`](@ref) that returns at least one column if `df` has at least one
+  column.
+
+# Examples
+
+```jldoctest
+julia> df = DataFrame(i=1:4, x=[1, 2, 1, 2])
+4×2 DataFrame
+ Row │ i      x
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      1
+   2 │     2      2
+   3 │     3      1
+   4 │     4      2
+
+julia> df = vcat(df, df)
+8×2 DataFrame
+ Row │ i      x
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      1
+   2 │     2      2
+   3 │     3      1
+   4 │     4      2
+   5 │     1      1
+   6 │     2      2
+   7 │     3      1
+   8 │     4      2
+
+julia> nonunique(df)
+8-element Vector{Bool}:
+ 0
+ 0
+ 0
+ 0
+ 1
+ 1
+ 1
+ 1
+
+julia> nonunique(df, keep=:last)
+8-element Vector{Bool}:
+ 1
+ 1
+ 1
+ 1
+ 0
+ 0
+ 0
+ 0
+
+julia> nonunique(df, 2)
+8-element Vector{Bool}:
+ 0
+ 0
+ 1
+ 1
+ 1
+ 1
+ 1
+ 1
+```
+"""
+function nonunique(df::AbstractDataFrame; keep::Symbol=:first)
+    if !(keep in (:first, :last, :only))
+        throw(ArgumentError("`keep` must be :first, :last, or :none"))
+    end
+    ncol(df) == 0 && return Bool[]
+    res = fill(true, nrow(df))
+    cols = ntuple(i -> df[!, i], ncol(df))
+    if keep == :first
+        # if we can take advantage of references pass groups to avoid generating hashes
+        rpa = refpool_and_array.(cols)
+        refpools = first.(rpa)
+        refarrays = last.(rpa)
+        if any(isnothing, refpools) || any(isnothing, refarrays)
+            ngroups, _, gslots, _ = row_group_slots!(cols, Val(true), nothing,
+                                                     false, nothing)
+            # unique rows are the first encountered group representatives,
+            # nonunique are everything else
+            cseen = 0
+            @inbounds for g_row in gslots
+                if g_row > 0
+                    res[g_row] = false
+                    # this check slows down the process when all rows are unique
+                    # but speeds up when we have duplicates
+                    cseen += 1
+                    cseen == ngroups && break
+                end
+            end
+        else
+            groups = Vector{Int}(undef, nrow(df))
+            ngroups = row_group_slots!(cols, refpools, refarrays,
+                                       Val(false), groups, false, false)[1]
+            seen = fill(false, ngroups)
+            cseen = 0
+            for i in 1:nrow(df)
+                g = groups[i]
+                if !seen[g]
+                    seen[g] = true
+                    res[i] = false
+                    cseen += 1
+                    cseen == ngroups && break
+                end
+            end
+        end
+    else
+        groups = Vector{Int}(undef, nrow(df))
+        ngroups = row_group_slots!(cols, Val(false), groups, false, nothing)[1]
+        if keep == :last
```

Comment:
1. For the case when we use ref-pools the performance is identical.
2. For the case when we cannot use ref-pools the performance is a bit worse (as we allocate `groups` and `seen` vectors, that are not needed otherwise), but I have checked that reversing (lazy) causes an even bigger drop in performance.

Here are benchmarks (without doing the reverse):
```
julia> df1 = DataFrame(x=string.(1:10^7));

julia> @btime nonunique($df1, keep=:first);
  505.621 ms (73 allocations: 213.84 MiB)

julia> @btime nonunique($df1, keep=:last);
  498.132 ms (77 allocations: 299.67 MiB)

julia> @btime nonunique($df1, keep=:nonduplicates);
  509.208 ms (77 allocations: 366.43 MiB)
```

(as you can see most impact is on memory allocation) - the time difference is so small that it is not visible.

Reply:
Actually I wasn't referring to the cost of allocating `groups`. What I'm concerned about are `refarray` objects, which should be faster for `:first` due to the special branch, than for `:last`.

Are you saying at your point 1. that `:first` and `:last` have the same performance for `refarray`? That would mean the special branch isn't useful?

Reply:
Here are complete benchmarks:
```
julia> df1 = DataFrame(x=1:10^7);

julia> df2 = string.(df1);

julia> @btime nonunique($df1, keep=:first);
  38.686 ms (73 allocations: 104.91 MiB)

julia> @btime nonunique($df1, keep=:last);
  38.791 ms (71 allocations: 104.91 MiB)

julia> @btime nonunique($df1, keep=:noduplicates);
  49.563 ms (71 allocations: 171.66 MiB)

julia> @btime nonunique($df2, keep=:first);
  508.361 ms (73 allocations: 213.84 MiB)

julia> @btime nonunique($df2, keep=:last);
  505.806 ms (77 allocations: 299.67 MiB)

julia> @btime nonunique($df2, keep=:noduplicates);
  527.655 ms (77 allocations: 366.43 MiB)

julia> # introduce duplicates

julia> repeat!(df1, 2);

julia> repeat!(df2, 2);

julia> @btime nonunique($df1, keep=:first);
  71.817 ms (73 allocations: 190.74 MiB)

julia> @btime nonunique($df1, keep=:last);
  72.180 ms (71 allocations: 190.74 MiB)

julia> @btime nonunique($df1, keep=:noduplicates);
  89.828 ms (71 allocations: 257.49 MiB)

julia> @btime nonunique($df2, keep=:first);
  942.053 ms (133 allocations: 427.67 MiB)

julia> @btime nonunique($df2, keep=:last);
  1.016 s (137 allocations: 589.80 MiB)

julia> @btime nonunique($df2, keep=:noduplicates);
  934.573 ms (137 allocations: 656.56 MiB)
```

In summary:
* if we can use refs `:first` and `:last` have the same performance (and this is expected)
* if we cannot use refs then `:first` uses a different algorithm that uses less memory and is a bit faster (but not by much) than `:last` (which uses the same algorithm as for refs); the major benefit is not using as much memory

Reply:
Actually I was confused as I forgot that `row_group_slots!` in the `:last` branch uses refarray if possible. I'll make suggestions to add comments about this.